packages/router/contracts/adapters/CurveAdapter.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.0;
+
+import '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
+import '../abstract/AdapterStorage.sol';
+import '../interfaces/IMarginlyRouter.sol';
+import './interfaces/ICurvePool.sol';
+
+contract CurveAdapter is AdapterStorage {
+  constructor(PoolInput[] memory pools) AdapterStorage(pools) {}
+
+  function swapExactInput(
+    address recipient,
+    address tokenIn,
+    address tokenOut,
+    uint256 amountIn,
+    uint256 minAmountOut,
+    bytes calldata data
+  ) external returns (uint256 amountOut) {
+    address poolAddress = getPoolSafe(tokenIn, tokenOut);
+    int128 tokenInIndex = ICurvePool(poolAddress).coins(0) == tokenIn
+      ? int128(0)
+      : int128(1);
+
+    // move all input tokens from router to this adapter
+    IMarginlyRouter(msg.sender).adapterCallback(address(this), amountIn, data);
+
+    amountOut = _swapExactInput(
+      poolAddress,
+      recipient,
+      tokenIn,
+      tokenInIndex,
+      amountIn,
+      minAmountOut
+    );
+
+    if (amountOut < minAmountOut) revert TooMuchRequested();
+  }
+
+  function swapExactOutput(
+    address recipient,
+    address tokenIn,
+    address tokenOut,
+    uint256 maxAmountIn,
+    uint256 amountOut,
+    bytes calldata data
+  ) external returns (uint256 amountIn) {
+    address poolAddress = getPoolSafe(tokenIn, tokenOut);
+
+    int128 tokenInIndex = ICurvePool(poolAddress).coins(0) == tokenIn
+      ? int128(0)
+      : int128(1);
+
+    // move all input tokens from router to this adapter
+    IMarginlyRouter(msg.sender).adapterCallback(address(this), maxAmountIn, data);
+
+    // swap all input tokens
+    uint256 actualAmountOut = _swapExactInput(
+      poolAddress,
+      address(this),
+      tokenIn,
+      tokenInIndex,
+      maxAmountIn,
+      amountOut
+    );
+
+    if (actualAmountOut < amountOut){
+      revert TooMuchRequested();
+    }
+
+    // move required part of actualAmountOut to recipient
+    SafeERC20.safeTransfer(IERC20(tokenOut), recipient, amountOut);
+
+    if (actualAmountOut == amountOut) {
+      amountIn = maxAmountIn;
+      return amountIn;
+    }
+
+    // amount of tokens that need to back swap
+    uint256 deltaAmountOut = actualAmountOut - amountOut;
+    uint256 amountOutMin = ICurvePool(poolAddress).get_dy(1 - tokenInIndex, tokenInIndex, deltaAmountOut);
+
+    // swap and move remained tokenIn directly to recipient
+    uint256 remainedTokenInAmount = _swapExactInput(
+      poolAddress,
+      recipient,
+      tokenOut,
+      1 - tokenInIndex,
+      deltaAmountOut,
+      amountOutMin
+    );
+
+    if (remainedTokenInAmount > maxAmountIn) {
```

Comment:
Is this even a realistic scenario? That would mean that curve pool both tokenIn and tokenOut balances decreased, which I'm sure pretty much impossible.

I would remove `if(){} else {}` here and leave just `amountIn = maxAmountIn - remainedTokenInAmount`